crates/storage/provider/src/transaction.rs

```diff
@@ -332,8 +504,466 @@ where
+                                                        * so it is same sa previous. */
+                });
+            }
+
+            let Some((_,block_transition)) = block_transition_iter.next() else { break};
+            // if block transition points to 1+next transition id it means that there is block
+            // changeset.
+            if block_transition == next_transition_id + 1 {
+                // assert last_transition_id == block_transition
+                if next_transition_id == next_changeset.0 {
+                    // take block changeset
+                    block_exec_res.block_changesets = next_changeset
+                        .1
+                        .into_iter()
+                        .map(|(address, (account, _))| (address, account))
+                        .collect();
+                    next_changeset = changeset_iter.next().unwrap_or_default();
+                }
+                next_transition_id += 1;
+            }
+            block_exec_results.push(block_exec_res)
+        }
+        Ok(block_exec_results)
+    }
+
+    /// Return range of blocks and its execution result
+    pub fn get_block_and_execution_range<const TAKE: bool>(
+        &self,
+        chain_spec: &ChainSpec,
+        range: impl RangeBounds<BlockNumber> + Clone,
+    ) -> Result<Vec<(SealedBlockWithSenders, ExecutionResult)>, TransactionError> {
+        if TAKE {
+            let from_transition = match range.start_bound() {
+                Bound::Included(n) => self.get_block_transition(n.saturating_sub(1))?,
+                Bound::Excluded(n) => self.get_block_transition(*n)?,
+                Bound::Unbounded => 0,
+            };
+            let to_transition = match range.end_bound() {
+                Bound::Included(n) => self.get_block_transition(*n)?,
+                Bound::Excluded(n) => self.get_block_transition(n.saturating_sub(1))?,
+                Bound::Unbounded => TransitionId::MAX,
+            };
+
+            let transition_range = from_transition..to_transition;
+            let zero = Address::zero();
+            let transition_storage_range =
+                (from_transition, zero).into()..(to_transition, zero).into();
+
+            self.unwind_account_hashing(transition_range.clone())?;
+            self.unwind_account_history_indices(transition_range.clone())?;
+            self.unwind_storage_hashing(transition_storage_range.clone())?;
+            self.unwind_storage_history_indices(transition_storage_range)?;
+
+            // merkle tree
+            {
+                let (tip_number, _) =
+                    self.cursor_read::<tables::CanonicalHeaders>()?.last()?.unwrap_or_default();
+                let current_root = self.get_header(tip_number)?.state_root;
+                let loader = DBTrieLoader::default();
+                let _block_root = loader.update_root(self, current_root, transition_range)?;
+            }
+            // check root
```

Comment:
do we need a TODO here for checking the state root?

Reply:
It is not needed as merkle tree of previous blocks is already checked when that block was included.

But added check for it, just for additional consistency check.